docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,74 @@
+
+This document is a draft of changes to ECMA-335 for the "runtime async" feature. When the feature is officially supported, it can be merged into the final ECMA-335 augments document.
+
+# Runtime-async
+
+Async is currently a feature implemented by various .NET languages as a compiler rewrite to support methods that can "yield" control back to their caller at specific "suspension" points. While effective, it's believed that implementation directly in the .NET runtime can provide improvements, especially in performance.
+
+## Spec modifications
+
+These are proposed modifications to the ECMA-335 specification for runtime-async.
+
+### I.8.4.5 Sync and Async Methods
+
+Methods may be either 'sync' or 'async'. Async method definitions are methods with a return type of `System.Threading.Task`, `System.Threading.ValueTask`, `System.Threading.Task<T>`, or `System.Threading.ValueTask<T>`  attributed with `[System.Runtimee.CompilerServices.RuntimeAsyncMethodAttribute]`. Async method definitions are only valid inside async-capable assemblies. An async-capable assembly is one which references a corlib containing an `abstract sealed class RuntimeFeature` with a `public const string` field member named `Async`, or a corelib meeting these requirements. The `RuntimeAsyncMethodAttribute` should only be applied to methods, and may be defined in any assembly. Inside async method bodies, certain methods are also invokable by a special signature encoding, described in [### I.8.6.1.5 Method signatures].
```

Comment:
What's the point of allowing `RuntimeAsyncMethodAttribute` to be defined in any assembly when it is only going to work on runtime-async enabled runtime?

Also, I am thinking whether it would be worth it to burn a MethodImplOptions bit for this.

Reply:
`MethodImplOptions` allows to save on metadata size if an attribute would be applicable frequently. But async methods are not very numerous. 

Are there other advantages that would be interesting here? 
Or just similarity with other things that are inputs to the JIT - like `AggressiveInlining`, `Synchronized`, .. ?

Reply:
> Are there other advantages that would be interesting here?

Eliminating cost of the attribute lookup in various places.

Reply:
> What's the point of allowing `RuntimeAsyncMethodAttribute` to be defined in any assembly when it is only going to work on runtime-async enabled runtime?

I asked Andy to document this because it will affect how Roslyn looks for the attribute, and some edge cases we will need to test. For our purposes, one of two things must be true:

* The attribute must be defined in the same assembly that defines `object` (practically, either System.Runtime.dll for the ref assembly, or System.Private.CoreLib for the runtime assembly), or
* The runtime must not care if the user defined their own copy of the attribute and Roslyn used it instead of the one defined in whatever assembly the "official" one lives in

Reply:
So, is it going to be an attribute or a new `MethodImpl` entry? Did we settle on this?

Reply:
I don't believe we settled on this, it should probably go in the open issues issue.

Reply:
For now I assume it is going to be a `MethodImpl` with the next not taken value - 1024.

Reply:
I've added an item to the design open issues. #109632

Reply:
Also added an item for runtime async methods in structs.